Improve Variables Tab Access and Bottom Panel Controls

abdul99ahad opened the pull request:
Closes #4516 

### Proposed Changes

<!--
Add relevant context (issue fixed or related to), 
a capture of the UI changes (if any) as well as 
steps to try out your changes.
--> 

### Variables Application Menu 

![Variables_Tab](https://github.com/user-attachments/assets/6ec94638-e244-4e50-8075-f2f5618362fc)


### Variables Tab Keyboard shortcut 
(For Mac OS): Press `Command + Shift + V`

![Variable Tab keyboard shortcut](https://github.com/user-attachments/assets/0c679f0c-e431-4292-927d-e19708f3d6a6)

### Variables Bottom Tab 

![Variable bottom tab](https://github.com/user-attachments/assets/604d698c-3540-48c9-a1c1-e5cd91aec91d)

### Checklist

To ensure you provided everything we need to look at your PR:

* [x] __Brief textual description__ of the changes present
* [x] __Visual demo__ attached
* [ ] __Steps to try out__ present, i.e. [using the `@bpmn-io/sr` tool](https://github.com/bpmn-io/sr)
* [x] Related issue linked via `Closes {LINK_TO_ISSUE}` or `Related to {LINK_TO_ISSUE}`

<!--
Thanks for creating this pull request! ❤️
-->

philippfromme commented:
We should call this _Toggle variables tab_

![image](https://github.com/user-attachments/assets/ac28f485-f33c-4204-8122-a9d7486ade9b)

I'd also add _Toggle problems tab_ now that we have the other entry.

barmac commented:
I'd suggest that we decide on another key for the shortcut. Ctrl/Cmd+V is the traditional paste shortcut, and in some editors a combination of Cmd+Shift+(Option)+V triggers "past and match formatting". Why don't we use a letter without strong OS link, e.g. U or M.

https://github.com/user-attachments/assets/a90c0cc8-c132-48ae-931f-6f96a5a4f4f7

nikku commented:
> I'd suggest that we decide on another key for the shortcut. Ctrl/Cmd+V is the traditional paste shortcut, and in some editors a combination of Cmd+Shift+(Option)+V triggers "past and match formatting". 

I subscribe to this. CC @lmbateman.

lmbateman commented:
I agree that we should stay away from Cmd-V variations. Looks like for many MacOS apps (but not the Desktop Modeler), Cmd-M minimizes the window. Cmd-U is a common shortcut for underline, but since that's a fairly specialized feature, I think Cmd-U will work, and Ctrl-U should be fine for Windows. (Should we add a Cmd-M shortcut?)

barmac commented:
BTW the recording I shared displays the vscode shortcuts. I think the Shift usage in that tool is wise as it allows to avoid conflicts. Learning from Cmd+P (print or toggle properties panel 🤡 ).

philippfromme commented:
MacOS tests are passing now. Will review.

philippfromme commented:
@abdul99ahad Could you adjust the PR according to my suggestion?

* add _Toggle bottom panel_ entry with <kbd>Ctrl+B</kbd> shortcut
* remove _Reset properties panel_ entry (was a workaround created at the time where we had issues with its size, no longer needed)
* decide whether to add individual entries for each tab

abdul99ahad commented:
> I'd imagine a toggle entry for the bottom panel (we could use Ctrl+B) and (optional) entries for the individual tabs.
> 
> ![image](https://private-user-images.githubusercontent.com/7633572/382713681-324326ba-4729-4bc3-a43f-c77d07153f44.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA3OTc5MDgsIm5iZiI6MTczMDc5NzYwOCwicGF0aCI6Ii83NjMzNTcyLzM4MjcxMzY4MS0zMjQzMjZiYS00NzI5LTRiYzMtYTQzZi1jNzdkMDcxNTNmNDQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMDVUMDkwNjQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2MzNmY3Y2RjZmY1ZmY3YjZmNDEzYzk2ZmNhNzFiMGNmZjlhM2VkYjkxNDU1YjZjZGI3MTlkNjkzMzcwNzFhOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.sPBA8R5ecU3E5a_Rl4GuBHmLDokafAjwFg4JWVIj76E)

I agree but the core issue here is the visibility of the Variables tab. One option could be to implement toggling for the entire bottom panel. Should we address this within the scope of this issue, or would it be better to open a new issue specifically for panel toggling and keep the focus here solely on the Variables panel? thoughts?

abdul99ahad commented:
> @abdul99ahad Could you adjust the PR according to my suggestion?
> 
> * add _Toggle bottom panel_ entry with Ctrl+B shortcut
> * remove _Reset properties panel_ entry (was a workaround created at the time where we had issues with its size, no longer needed)
> * decide whether to add individual entries for each tab

Okay. Regarding point 3, I think for now, we can add variables tab as individual entry and see in future if users also want Output and Problems tab in application menu.

nikku commented:
> Regarding point 3, I think for now, we can add variables tab as individual entry and see in future if users also want Output and Problems tab in application menu.

I'd say: Don't add a toggle for variables panel for now, but in the future automatically open variables tab from various places. Much more important that this is properly wired, i.e. with the FEEL editors.

philippfromme commented:
Additionally, adding the individual entries including disabled states is a bit more work since opening a tab doesn't trigger a menu update at the moment. So I'd say let's keep it for later or just don't add them if no one ever asks.

nikku commented on client/src/app/__tests__/AppSpec.js:
```diff
@@ -2085,6 +2085,60 @@ describe('<App>', function() {
+          });
+
+          // when
+          app.triggerAction('open-bottom-panel');
```
```suggestion
          app.triggerAction('toggle-bottom-panel');
```

nikku replied:
Let's rename to `toggle`, as toggle is what it is.

nikku commented on client/src/app/tabs/bpmn/BpmnEditor.js:
```diff
@@ -466,7 +466,8 @@ export class BpmnEditor extends CachedComponent {
       undo: commandStack.canUndo(),
-      zoom: true
+      zoom: true,
+      bottomPanel: true
```
What is the purpose of this variable? Why do we explicitly need to configure it? I assume that the _bottom panel_ is always available, one way or the other.

abdul99ahad replied:
Agreed. Since the bottom panel is universally accessible, configuring this variable explicitly seems unnecessary.

nikku replied:
Why would the `bottomPanel` entry point still be in here? We want it to be toggleable from _any_ view.

nikku replied:
I.e. also from the welcome page.

abdul99ahad commented:
> Let's restore the previous bottom panel state (if applicable), once I toggle it open (again). This is standard behavior, and ensures I can temporarily close things, re-open if needed, and end up in the previous setup.
> 
> Right now I always end up on the `Output` tab, which is not desirable.
> 
> ![capture wbkSyL_optimized](https://private-user-images.githubusercontent.com/58601/383148635-b56b6020-f6db-40ec-97a0-2c36c3cee89d.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEzMjI1MjQsIm5iZiI6MTczMTMyMjIyNCwicGF0aCI6Ii81ODYwMS8zODMxNDg2MzUtYjU2YjYwMjAtZjZkYi00MGVjLTk3YTAtMmMzNmMzY2VlODlkLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTExVDEwNTAyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA5MDUyNTZlYTFkNGUyZTAwNzBmODc3ZjJlMmU4ZmZhODY1NDI2YzIzNDM4MDlmMTAzZDRlNDk2YTUxNTY0ZjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.T17NjtP_j5BH1hIhUR5K2oZz_QkZYX8mXtflpjlVnEQ) [ ![capture wbkSyL_optimized](https://private-user-images.githubusercontent.com/58601/383148635-b56b6020-f6db-40ec-97a0-2c36c3cee89d.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEzMjI1MjQsIm5iZiI6MTczMTMyMjIyNCwicGF0aCI6Ii81ODYwMS8zODMxNDg2MzUtYjU2YjYwMjAtZjZkYi00MGVjLTk3YTAtMmMzNmMzY2VlODlkLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTExVDEwNTAyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA5MDUyNTZlYTFkNGUyZTAwNzBmODc3ZjJlMmU4ZmZhODY1NDI2YzIzNDM4MDlmMTAzZDRlNDk2YTUxNTY0ZjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.T17NjtP_j5BH1hIhUR5K2oZz_QkZYX8mXtflpjlVnEQ) ](https://private-user-images.githubusercontent.com/58601/383148635-b56b6020-f6db-40ec-97a0-2c36c3cee89d.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEzMjI1MjQsIm5iZiI6MTczMTMyMjIyNCwicGF0aCI6Ii81ODYwMS8zODMxNDg2MzUtYjU2YjYwMjAtZjZkYi00MGVjLTk3YTAtMmMzNmMzY2VlODlkLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTExVDEwNTAyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA5MDUyNTZlYTFkNGUyZTAwNzBmODc3ZjJlMmU4ZmZhODY1NDI2YzIzNDM4MDlmMTAzZDRlNDk2YTUxNTY0ZjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.T17NjtP_j5BH1hIhUR5K2oZz_QkZYX8mXtflpjlVnEQ) [ ](https://private-user-images.githubusercontent.com/58601/383148635-b56b6020-f6db-40ec-97a0-2c36c3cee89d.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEzMjI1MjQsIm5iZiI6MTczMTMyMjIyNCwicGF0aCI6Ii81ODYwMS8zODMxNDg2MzUtYjU2YjYwMjAtZjZkYi00MGVjLTk3YTAtMmMzNmMzY2VlODlkLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTExVDEwNTAyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA5MDUyNTZlYTFkNGUyZTAwNzBmODc3ZjJlMmU4ZmZhODY1NDI2YzIzNDM4MDlmMTAzZDRlNDk2YTUxNTY0ZjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.T17NjtP_j5BH1hIhUR5K2oZz_QkZYX8mXtflpjlVnEQ)

Done. Now, the bottom panel maintained state and re-open the same tab that was previously opened.

nikku commented on client/src/app/App.js:
```diff
@@ -1946,6 +1946,11 @@ export class App extends PureComponent {
       return this.emitWithTab(type, activeTab, payload);
     }
 
+    if (action === 'toggleBottomPanel') {
```
We already have the action `close-panel`, which closes the bottom panel.

I'd suggest we rename that action to `close-bottom-panel` and this particular action to `toggle-bottom-panel`. `camelCase` is not the pattern we use.

```suggestion
    if (action === 'toggle-bottom-panel') {
```

abdul99ahad replied:
I originally used `toggle-button-panel`, but since all actions in the file are in `camelCase`, I updated it to match that style.
<img width="457" alt="image" src="https://github.com/user-attachments/assets/9c4a93f0-b640-4913-997b-cdf54193222a">

nikku commented on client/src/app/__tests__/AppSpec.js:
```diff
@@ -2114,6 +2114,90 @@ describe('<App>', function() {
 
     });
 
+    describe('bottom-panel', function() {
```
Mind our usual layout:

* Two empty lines between tests (`it`), two empty lines between tests + setup or consecutive `describe` blocks.

abdul99ahad replied:
Alright. From what I have seen, this layout behavior is inconsistent throughout the testcases 

cf. 
<img width="800" alt="image" src="https://github.com/user-attachments/assets/f6402df8-3bc5-420a-b092-66132bac02c2">

nikku replied:
If that is the case, then let's consistently apply new baseline ([boy scout rule](https://eugenhoble.medium.com/basic-programming-principles-the-boy-scout-rule-5283af4e15c0)).

nikku commented on client/src/app/tabs/cloud-bpmn/BpmnEditor.js:
```diff
@@ -432,7 +432,7 @@ export class BpmnEditor extends CachedComponent {
       spaceTool: !inputActive,
       undo: commandStack.canUndo(),
-      zoom: true
+      zoom: true,
```
Remove.

nikku reviewed (changes requested):
Ok, I think we're close to the finish line.

Generally, consider if your code is consistent with what we currently have. Let's follow existing patterns on a best effort basis.

abdul99ahad commented:
All changes are incorporated, and this PR is ready for review. Here’s a summary of the updates:

- Variables Tab: Now accessible via the bottom tab button.
- Toggle Bottom Panel:
  - Added to the application menu.
  - The panel now preserves context when opened/closed.
  - Accessible with the keyboard shortcut Command + B (MacOS) or Ctrl + B (Windows).
- Removed: The “Reset Properties” option from the application menu.

nikku commented:
@abdul99ahad Regarding https://github.com/camunda/camunda-modeler/pull/4638/commits/aa5d66d9043355041d1eb7ee2ae84242320e5c5c: The usual pattern is two empty lines separating two specs, separating specs and setup, separating two describes.

We don't add two empty lines between describe and child spec or describe, so this is the standard way of formatting you'll find across 99% of our tests:

```javascript
describe('foo', function() {

  beforeEach(...);
  
  beforeEach(...);
  
  
  it('first test');
  

  it('second test');
  
  
  describe('nested describe', function() {
    
    describe('deeply nested describe', function() {
      
      it('nested it');
    });
    
  });
  
  
  describe('another nested describe', ...);
  
});
```

nikku reviewed (changes requested):
Beyond a minor style comment (https://github.com/camunda/camunda-modeler/pull/4638#issuecomment-2475560159) this still seems to not work as expected:

* What I'd expect is that in _any_ view I'm able to toggle the bottom tabs open and close.
* What I see is that this PR manually registers the action for a subset of our views, so it is not available in forms, welcome page, etc.

abdul99ahad commented:
> @abdul99ahad Regarding [aa5d66d](https://github.com/camunda/camunda-modeler/commit/aa5d66d9043355041d1eb7ee2ae84242320e5c5c): The usual pattern is two empty lines separating two specs, separating specs and setup, separating two describes.
> 
> We don't add two empty lines between describe and child spec or describe, so this is the standard way of formatting you'll find across 99% of our tests:
> 
> ```js
> describe('foo', function() {
> 
>   beforeEach(...);
>   
>   beforeEach(...);
>   
>   
>   it('first test');
>   
> 
>   it('second test');
>   
>   
>   describe('nested describe', function() {
>     
>     describe('deeply nested describe', function() {
>       
>       it('nested it');
>       
>     });
>     
>   });
>   
>   
>   describe('another nested describe', ...);
>   
> });
> ```

Understood. If that's the convention, then I think the test cases are perfectly aligned with the convention.

abdul99ahad commented:
> Beyond a minor style comment ([#4638 (comment)](https://github.com/camunda/camunda-modeler/pull/4638#issuecomment-2475560159)) this still seems to not work as expected:
> 
> * What I'd expect is that in _any_ view I'm able to toggle the bottom tabs open and close.
> * What I see is that this PR manually registers the action for a subset of our views, so it is not available in forms, welcome page, etc.

Agreed. I assumed it should function like the properties panel, but since the bottom panel is more global, it makes sense to make it universally available. I've implemented the necessary changes, and the bottom panel is now accessible throughout the modeler.

nikku commented:
Ok, this works properly now, across all file types. One issue that remains is that we don't properly fall back to another "known tab" when we switch to a view where a particular tab is not present: 

![capture R4Me67_optimized](https://github.com/user-attachments/assets/434f4797-3ae0-4c60-9591-8f0e08d8171d)

This issue, however, can already be reproduced in released versions of the modeler, so not a regression.

nikku commented:
I think it is a good solution we ended up building here :clap:

nikku commented:
Please add an entry to the CHANGELOG, linking to the issue this closes.

nikku commented:
Merging with some minor squashes.

philippfromme commented:
Question: Did we agree to move the menu entry? I would expect it to be grouped with the `Toggle properties panel` entry.

![image](https://github.com/user-attachments/assets/57013e96-db28-414c-a365-c6041b2b5a24)

abdul99ahad commented:
> Question: Did we agree to move the menu entry? I would expect it to be grouped with the `Toggle properties panel` entry.
> 
> ![image](https://private-user-images.githubusercontent.com/7633572/388948320-57013e96-db28-414c-a365-c6041b2b5a24.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIyODMyMzUsIm5iZiI6MTczMjI4MjkzNSwicGF0aCI6Ii83NjMzNTcyLzM4ODk0ODMyMC01NzAxM2U5Ni1kYjI4LTQxNGMtYTM2NS1jNjA0MWIyYjVhMjQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjJUMTM0MjE1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2MwMjA2YWQ0NWZiOWZlZmI0MDYxNzM2NGEyNzlhZTQwZTI3YTFiNDFhNTc3NzBmMjkzOTgxODA5ZTEyNTc5ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.72vtDTERX7bVM2KqZ91kLZ_jeT4tLO0jaWXdLrzJzyo)

Initially, the "Toggle Bottom Panel" option was grouped with `Toggle Properties Panel` because both were associated with specific views like BPMN and DMN diagrams. However, since the Bottom Panel is universally available across the entire application, its corresponding menu button needs to be accessible everywhere. Therefore, we’ve decided to place it alongside other universally applicable options in the application menu.

philippfromme commented:
Okay, understood. 😃
